### ISSUE TITLE
Checkbox component to work well with useForm

### ISSUE THREAD

**Issue by nk922** [MetroStar/comet#209]
**Is your feature request related to a problem? Please describe.**
Yes.  If I use the useForm hook I am not able to change the checked property of the checkbox as part of an event handler.  It always stays  to what I set as part of the first render.

**Describe the solution you'd like**
Have the checked property bound to the form value just like TextInput.

**Describe alternatives you've considered**
I have considered attaching a ref to the Checkbox but that evaluates to null since Checkbox does not seem to be forwarding a ref.

**Additional context**

const {
    control,
    reset,
    handleSubmit,
    formState: { errors, isDirty },
  } = useForm<SettingsFormInput>({
    defaultValues: SettingsFormDefaults,
  });

const checkbox1Ref = useRef<HTMLInputElement>(null);

function handleRestoreToDefaults() {
     **//this is the line that's able to update values for a textbox but not the checkboxes**     
     reset(getDefaultsFromServiceResponse(getDefaultSettings()));
    //item.useNewKnowledgeBaseForm = getDefaultSettings().useNewKnowledgeBaseForm;
**//   checkbox1Ref evaluates to null**
if (checkbox1Ref.current) {
      checkbox1Ref.current.checked = false;
    }
  }

<FormGroup>
                <Controller
                  name="useNewKnowledgeBaseForm"
                  control={control}
                  // eslint-disable-next-line @typescript-eslint/no-unused-vars
                  render={({ field: { ref: _, ...field } }) => (
                    <Checkbox
                      {...field}
                      label="Use New Knowledge Base Form"
                      id="useNewKnowledgeBaseForm"
                      checked={item.useNewKnowledgeBaseForm}
                      isTile={false}
                      value={'Y'}
                      ref={checkbox1Ref}
                    />
                  )}
                />

**Comment by jbouder** [comment]
I believe `defaultValue` is not a valid attribute for boolean type fields, such as a checkbox or radio button. Have you tried providing those defaults as `defaultChecked` or `checked` within your useForm object?

**Comment by nk922** [comment]
I am currently using checked attribute on the Checkbox component.  Changing that to defaultChecked does not have any impact.

Here is a snippet of what my form data interface looks like for a form with one textbox and two checkboxes:

export interface SettingsFormInput {
  pageSize: number;
  //askConfigId: string;
  suppressChatCreatedAssets: boolean;
  useNewKnowledgeBaseForm: boolean;
}

While the pageSize which gets automatically bound to a textbox field; the two boolean fields don't get bound to the checkbox automatically and I manually have to initialize them via the checked or defaultChecked attribute on the Checkbox component like below:
`<Controller
                  name="suppressChatCreatedAssets"
                  control={control}
                  // eslint-disable-next-line @typescript-eslint/no-unused-vars
                  render={({ field: { ref: _, ...field } }) => (
                    <Checkbox
                      {...field}
                      label="Suppress Chat Created Assets"
                      id="suppressChatCreatedAssets"
                      defaultChecked={item.suppressChatCreatedAssets}
                      isTile={false}
                      value={'Y'}
                    />
                  )}
                />`
Issue then becomes how do I check/uncheck the checkbox programmatically?  Textbox responds well to reset method from useForm but the checkboxes don't.  Even if I change the values on the item object for item.suppressChatCreatedAssets the checbox value does not change, if I click away from that component and come back it shows the update but it's too late; I want it to change as part of the event handler itself.

**Comment by jbouder** [comment]
> I am currently using checked attribute on the Checkbox component. Changing that to defaultChecked does not have any impact.
> 
> Here is a snippet of what my form data interface looks like for a form with one textbox and two checkboxes:
> 
> export interface SettingsFormInput { pageSize: number; //askConfigId: string; suppressChatCreatedAssets: boolean; useNewKnowledgeBaseForm: boolean; }
> 
> While the pageSize which gets automatically bound to a textbox field; the two boolean fields don't get bound to the checkbox automatically and I manually have to initialize them via the checked or defaultChecked attribute on the Checkbox component like below: `<Controller name="suppressChatCreatedAssets" control={control} // eslint-disable-next-line @typescript-eslint/no-unused-vars render={({ field: { ref: _, ...field } }) => ( <Checkbox {...field} label="Suppress Chat Created Assets" id="suppressChatCreatedAssets" defaultChecked={item.suppressChatCreatedAssets} isTile={false} value={'Y'} /> )} />` Issue then becomes how do I check/uncheck the checkbox programmatically? Textbox responds well to reset method from useForm but the checkboxes don't. Even if I change the values on the item object for item.suppressChatCreatedAssets the checbox value does not change, if I click away from that component and come back it shows the update but it's too late; I want it to change as part of the event handler itself.

Good to know. I'll do some testing and will get back to you soon.

**Comment by nk922** [comment]
not seeing any difference.  do I need to make any corresponding changes to how my form input is defined, checkbox markup, or how I reset the form to default values using a specific instance of the form input?

**Comment by jbouder** [comment]
> input?

You’ll need to use the `defaultChecked` and `checked` props respectively .

**Comment by nk922** [comment]
sorry, I am not following.  Could you please share a snippet of the modified code I should be using highlighting the changes?

**Comment by jbouder** [comment]
> sorry, I am not following. Could you please share a snippet of the modified code I should be using highlighting the changes?

Here’s an example story with a basic setup: https://github.com/MetroStar/comet/pull/211/files#diff-18480eb93581320063df322dfcfea457b7fec771b5243e51ca887f6b457ad6efR48

Let me know if that is enough. If not, I can get a more in depth example up

**Comment by nk922** [comment]
ok, so this is recommending using only the **defaultChecked** prop on the checkbox, right?

How do I check/uncheck the checkbox at runtime in response to a user clicking a button in an eventhandler?

**Comment by jbouder** [comment]
> ok, so this is recommending using only the **defaultChecked** prop on the checkbox, right?
> 
> How do I check/uncheck the checkbox at runtime in response to a user clicking a button in an eventhandler?

This change will also be applied to the radio button after I verify functionality. 

That said, you should be able to use the `checked` prop for realtime state. You would just need to make sure it’s tracked with some state like a hook (useState or useForm)

**Comment by nk922** [comment]
can you please share an example with the useForm managing the checkbox state at runtime?  I am trying to avoid using useState along with the useForm just for the checkboxes.

and how do I use the checked prop at runtime in my event handler code?

**Comment by jbouder** [comment]
> can you please share an example with the useForm managing the checkbox state at runtime? I am trying to avoid using useState along with the useForm just for the checkboxes.
> 
> and how do I use the checked prop at runtime in my event handler code?

I'll try to get something up today or tomorrow.

**Comment by jbouder** [comment]
Everything appears to be working as I would expect. I created a dummy PR here: https://github.com/jbouder/comet-starter/pull/1.

The PR includes the beta version as well as a basic implementation of a checkbox being controlled by useForm. The screenshots show that the value is being tracked in realtime be logging to the console, as well as preventing the form from submitting.

**Comment by nk922** [comment]
Couple of observations/questions/extension use case(s):

1. **accept** is defined as a string and as a boolean on the FormInput.  Is this by design?
2. useForm on the sign-in.tsx deliberately skips the **accept** property.  Is this by design?
3. on the actual Checkbox component defaultChecked is used.  Is it possible to set it to true/false initial value based on some runtime condition at the time of initial render instead of just leaving it like this (which I assume makes it evaluate to true)?
4. If I then add a new button next to Cancel named "Reset terms and conditions" how can I programmatically uncheck the checkbox and potentially set other form fields (like username) to custom values?  is there an API like reset or setValue from useForm which will support setting the checkbox to a custom state (checked/unchecked)?

**Comment by jbouder** [comment]
> Couple of observations/questions/extension use case(s):
> 
> 1. **accept** is defined as a string and as a boolean on the FormInput.  Is this by design?
> 2. useForm on the sign-in.tsx deliberately skips the **accept** property.  Is this by design?
> 3. on the actual Checkbox component defaultChecked is used.  Is it possible to set it to true/false initial value based on some runtime condition at the time of initial render instead of just leaving it like this (which I assume makes it evaluate to true)?
> 4. If I then add a new button next to Cancel named "Reset terms and conditions" how can I programmatically uncheck the checkbox and potentially set other form fields (like username) to custom values?  is there an API like reset or setValue from useForm which will support setting the checkbox to a custom state (checked/unchecked)?  This is the use case I am struggling with the most.

1. I'm only seeing accept defined as a string, which is because the "value" is a string. And it appears that react-hook-forms expects it that way as well.
2. I actually was seeing a console error related to how I was setting the defaultChecked value, so went this route, and seems to be better now.
3. You should definitely be able to set defaultChecked to some dynamic value. Just keep in mind that defaultChecked will only apply when the component is initially loaded, after that you'll need to handle changes through the checked prop.
4. I'm no expert on react-hook-forms, but was able to getting it working by setting checked based on the value.

Please take a look at the PR again, I made some updates and I believe I now have what you're looking for :)

**Comment by nk922** [comment]
Calling reset({ accept: 'true' }); has no effect.  If the user manually unchecks the checkbox this code does not check it back.

**Comment by jbouder** [comment]
> Calling reset({ accept: 'true' }); has no effect. If the user manually unchecks the checkbox this code does not check it back.

The example I provided is just a basic example illustrating both setting a default value (defaultChecked) as well as providing a way to update the checked/unchecked value of the checkbox component. I'm using reset as an example. And my implementation is just reseting to checked, so you only notice the clear button doing anything if you uncheck first. See the below recording.

https://github.com/user-attachments/assets/e56990b1-d7ad-4d76-a003-6803b4b938c6

All of that said, my example again is a basic example making use of the Comet Checkbox as well as React Hook Form. It appears to be working as I would expect, I think you just need to tweak your code a bit for your use case. 

Please let me know if that makes sense.

**Comment by nk922** [comment]
Interesting, when I run this sample console does print true from the below code but the checkbox remains unchecked for the user visually - I tried multiple browsers:

const isChecked = watch('accept');
 useEffect(() => {
    **console.log(isChecked);**
  }, [isChecked]);

I'll incorporate the changes into my application as well and see if it's a different behavior there.

**Comment by nk922** [comment]
The sample you shared is working for me now and it's also working in my actual application after I incorporated the changes.

However, on another note, I have run into some issues with the modal component.  Whenever user clicks on continue I am having issues with closing the modal using the below line of code:

modal.toggleModal.call(modal, new KeyboardEvent('keydown', {}));

It used to work just fine with the version 2.0.1 of metrostar/comet-uswds.  We are currently using 2.0.1 and upgrading to 2.3.0-beta.0 seems to break the closing the modal after user clicks on Continue and we do some processing.

The following snippet does not work, I get a compilation error:

/* eslint-disable-next-line */
//modal.toggleModal.call(modal, false); // Manually toggle modal after work complete

The TypeScript error is: Argument of type 'boolean' is not assignable to parameter of type 'KeyboardEvent'.ts(2345)

Could you please review this modal popup behavior?  If you give me the different versions of the library I might be able to confirm which version starts to cause this behavior.

We won't be able upgrade the version until modal popup issues are resolved.

Thanks,
Narayan

**Comment by jbouder** [comment]
> The sample you shared is working for me now and it's also working in my actual application after I incorporated the changes.
> 
> However, on another note, I have run into some issues with the modal component. Whenever user clicks on continue I am having issues with closing the modal using the below line of code:
> 
> modal.toggleModal.call(modal, new KeyboardEvent('keydown', {}));
> 
> It used to work just fine with the version 2.0.1 of metrostar/comet-uswds. We are currently using 2.0.1 and upgrading to 2.3.0-beta.0 seems to break the closing the modal after user clicks on Continue and we do some processing.
> 
> The following snippet does not work, I get a compilation error:
> 
> /* eslint-disable-next-line */ //modal.toggleModal.call(modal, false); // Manually toggle modal after work complete
> 
> The TypeScript error is: Argument of type 'boolean' is not assignable to parameter of type 'KeyboardEvent'.ts(2345)
> 
> Could you please review this modal popup behavior? If you give me the different versions of the library I might be able to confirm which version starts to cause this behavior.
> 
> We won't be able upgrade the version until modal popup issues are resolved.
> 
> Thanks, Narayan

I went ahead and reproduced the custom example in the Comet Storybook example in the Comet Starter (using the latest release of comet-uswds), and it seems to be working as expected. Here is the basic example: https://github.com/jbouder/comet-starter/pull/2

**Comment by nk922** [comment]
After upgrading the @uswds/uswds to 3.8.1 (I was using 3.7.0 earlier) I am able to get past the modal popup issue and validate the issue at runtime.  However, I am unable to build my docker image for deployment due to the following error...

------
 > [build 6/6] RUN npm run build:
0.465
0.465 > comet-starter@1.0.0 build
0.465 > tsc && vite build
0.465
3.130 src/components/delete/delete.tsx(**72,35**): error TS2345: Argument of type 'boolean' is not assignable to parameter of type 'KeyboardEvent'.

Line #72 has the following code:
modal.toggleModal.call(modal, false); // Manually toggle modal after work complete

VS Code shows a squiggly red underline below false but lets me run the code via npm run dev but I still gets stuck when building the docker image via **docker build -t aiapp .**

**Comment by jbouder** [comment]
> After upgrading the @uswds/uswds to 3.8.1 (I was using 3.7.0 earlier) I am able to get past the modal popup issue and validate the issue at runtime. However, I am unable to build my docker image for deployment due to the following error...
> 
> > [build 6/6] RUN npm run build:
> > 0.465
> > 0.465 > comet-starter@1.0.0 build
> > 0.465 > tsc && vite build
> > 0.465
> > 3.130 src/components/delete/delete.tsx(**72,35**): error TS2345: Argument of type 'boolean' is not assignable to parameter of type 'KeyboardEvent'.
> 
> Line #72 has the following code: modal.toggleModal.call(modal, false); // Manually toggle modal after work complete
> 
> VS Code shows a squiggly red underline below false but lets me run the code via npm run dev but I still gets stuck when building the docker image via **docker build -t aiapp .**

Probably not related, but I did actually have to fix a tsconfig issue (which popped up when I replaced jest with vitest), which was breaking my build. That change was merged into main and the PR listed here, in case you wanted to look: https://github.com/MetroStar/comet-starter/commit/b016c20ffdd03418cd3e0bb5ddf146334b8248e7#diff-b55cdbef4907b7045f32cc5360d48d262cca5f94062e353089f189f4460039e0L39

As for your modal build issue. I'm guessing the error might be related to how you're declaring the uswds modal module. Looking at the source for the uswds modal, it does look like it should receive an event as a parameter, however you can also provide a boolean like we are here. This is based on the following use within the USWDS component: https://github.com/uswds/uswds/blob/369cf3fdf9abf721a357e864a0f6adbf667ba862/packages/usa-modal/src/index.js#L39. 

On the above I'm adding the usa-modal module to my types.d.ts file here, which is pretty generic and might explain why TypeScript is complaining and mine isn't, if you're doing this differently. Here is where I'm doing that: https://github.com/jbouder/comet-starter/pull/2/files#diff-23c7d6d050a365564a07eeef435d1d9fc8ea94230a751483afe643435cc9647dR2

**Comment by nk922** [comment]
Thanks!

I already have "include": ["src"] in my tsconfig.json file (no "vite.config.ts").

Adding declare module '@uswds/uswds/js/usa-modal'; to types.d.ts did the trick to resolve the build error.  Thanks a lot!

**Comment by jbouder** [comment]
> Thanks!
> 
> I already have "include": ["src"] in my tsconfig.json file (no "vite.config.ts").
> 
> Adding declare module '@uswds/uswds/js/usa-modal'; to types.d.ts did the trick to resolve the build error. Thanks a lot!

Thats great to hear! Glad we were able to figure it out. I'll get a new release out shortly, just waiting on a review. Will post here once thats ready.

**Comment by jbouder** [comment]
@nk922 , release 2.3.0 has been published. Thanks for posting the issue!

**Comment by nk922** [comment]
I think I have run into the RadioButtonGroup control with this package upgrade of @metrostar/comet-uswds and @uswds/uswds.  Here is my basic setup of the RadioButtonGroup.

export const textSourceRadioButtonOptions = [
  { label: 'Manually Enter Text', value: 'manual', checked: false },
  { label: 'Retrieve Text from URI', value: 'uri', checked: false },
  { label: 'Upload File for Text', value: 'file', checked: false },
];

const handleTextSourceChange: ChangeEventHandler<HTMLInputElement> = (
    event,
  ) => {
    console.log('event.target.value: ', event.target.value);
    setTextSource(event.target.value);
    setValue('textSourceValidationHelper', **event.target.value**, {
      shouldValidate: true,
      shouldDirty: true,
    });
  };

<FormGroup>
                <Label htmlFor="textSource">Select a text source</Label>
                <Controller
                  name="textSource"
                  control={control}
                  rules={REQUIRED_FORM_FIELDS_RULES}
                  // eslint-disable-next-line @typescript-eslint/no-unused-vars
                  render={({ field: { ref: _, ...field } }) => (
                    <RadioButtonGroup
                      {...field}
                      id="textSource"
                      name="textSource"
                      data={textSourceRadioButtonOptions}
                      onChange={handleTextSourceChange}
                    />
                  )}
                />
                {/* currently the validations for the radiobuttongroup don't work out of the box */}
                {errors.textSource?.message && (
                  <ErrorMessages errors={[errors.textSource.message]} />
                )}
                <Controller
                  name="textSourceValidationHelper"
                  control={control}
                  rules={REQUIRED_FORM_FIELDS_RULES}
                  // eslint-disable-next-line @typescript-eslint/no-unused-vars
                  render={({ field: { ref: _, ...field } }) => (
                    <input
                      {...field}
                      type={SELECTED_TAGS_CONTROL_TYPE}
                      id="textSourceValidationHelper"
                      name="textSourceValidationHelper"
                    ></input>
                  )}
                />
                {errors.textSourceValidationHelper?.message && (
                  <ErrorMessages
                    errors={[errors.textSourceValidationHelper.message]}
                  />
                )}
              </FormGroup>

Earlier when I selected one of the radio buttons I used to get the value of the button from event.target.value in the event handler handleTextSourceChange.  Highlighting of the selected radio button was also correct as shown below.

![image](https://github.com/user-attachments/assets/d54626d9-44ba-413a-b6e6-885f19f61e27)

![image](https://github.com/user-attachments/assets/66b6601d-00b1-459c-8445-56350ddbde97)

Now when I select a radio button the feedback to the user is not given.  Also, the value that I get from event.target.value is just 'on' no matter which option I choose.  e.g.

![image](https://github.com/user-attachments/assets/0e138128-88bb-40a0-897c-2f0d7d657791)

The radio button that I click on receives the focus and as soon as focus away it seems none of them are checked.

![image](https://github.com/user-attachments/assets/25cceaa9-541c-46b8-9718-ae28085fec03)

**Comment by jbouder** [comment]
In reference to using the radio group and being able to get a value out as the group changes, I was able to get an implementation up here: https://github.com/jbouder/comet-starter/compare/main...jbouder:comet-starter:radio-group-test. As far as I can tell, it seems to be working as expected. Can you review and compare with your implementation?

**Comment by nk922** [comment]
when I run this locally I get 'on' written into the console from target.value

**Comment by jbouder** [comment]
> when I run this locally I get 'on' written into the console from target.value not 'manual', 'uri', or 'text' as expected

Interesting. I'm not seeing that. Here is what I see.
<img width="1728" alt="Screenshot 2024-08-06 at 6 20 24 PM" src="https://github.com/user-attachments/assets/d5101d60-0a7c-4629-bc67-800c8528d39e">

I might suggest ensuring you're on the newest version (2.4.0), and reviewing your specific implementation against my example.

**Comment by nk922** [comment]
I got to the right branch and it seems to be working fine.  Also, for my code base upgrading to 2.4.0 seems to resolve the issue.

Thanks a lot!

I'll revert back in case I see any additional issues but seems like I should be all set.

**Comment by jbouder** [comment]
> I got to the right branch and it seems to be working fine. Also, for my code base upgrading to 2.4.0 seems to resolve the issue.
> 
> Thanks a lot!
> 
> I'll revert back in case I see any additional issues but seems like I should be all set.

Excellent! Sorry about any confusion on the recent updates. And thanks again for keeping me in the loop!
